python/ray/serve/tests/test_callback.py

```diff
@@ -0,0 +1,205 @@
+    with pytest.raises(ValueError, match="cannot be imported"):
+        call_function_from_import_path("not_exist")
+
+    # raise exception when providing non callable object
+    with pytest.raises(TypeError, match="is not callable"):
+        call_function_from_import_path(
+            "ray.serve.tests.test_callback.NOT_CALLABLE_OBJECT"
+        )
+
+
+@pytest.mark.parametrize(
+    "ray_instance",
+    [
+        {
+            "RAY_SERVE_CONTROLLER_CALLBACK_IMPORT_PATH": ADD_LOGGER_IMPORT_PATH,
+            "RAY_SERVE_HTTP_PROXY_CALLBACK_IMPORT_PATH": ADD_MIDDLEWARE_IMPORT_PATH,
+        },
+    ],
+    indirect=True,
+)
+def test_callback(ray_instance, capsys):
+    """Test callback function works in http proxy and controller"""
+
+    @serve.deployment
+    class Model:
+        def __call__(self, request: starlette.requests.Request):
+            headers = request.scope.get("headers")
+            for k, v in headers:
+                if k == b"custom_header_key":
+                    return v
+            return "Not found custom headers"
+
+    serve.run(Model.bind())
+    resp = requests.get("http://localhost:8000/")
+    assert resp.text == "custom_header_value"
+
+    captured = capsys.readouterr()
+    assert "MyCustom message: hello" in captured.err
+
+
+@pytest.mark.parametrize(
+    "ray_instance",
+    [
+        {
+            "RAY_SERVE_CONTROLLER_CALLBACK_IMPORT_PATH": "not_exist",
+            "RAY_SERVE_HTTP_PROXY_CALLBACK_IMPORT_PATH": RAISE_ERROR_IMPORT_PATH,
+        },
+    ],
+    indirect=True,
+)
+def test_callback_fail(ray_instance):
+    """Test actor call call_function_from_import_path rasing exception.
+
+    Actor will fail to be started and further call will raise RayActorError.
+    """
+
+    actor_def = ray.serve._private.http_proxy.HTTPProxyActor
+    handle = actor_def.remote(
+        host="http_proxy",
+        port=123,
+        root_path="/",
+        controller_name="controller",
+        node_ip_address="127.0.0.1",
+    )
+    with pytest.raises(RayActorError, match="this is from raise_error_callback"):
+        ray.get(handle.ready.remote())
+
+    actor_def = ray.serve.controller.ServeController
+    handle = actor_def.remote(
+        "controller",
+        http_config={},
+        head_node_id="123",
+    )
+    with pytest.raises(RayActorError, match="cannot be imported"):
+        ray.get(handle.check_alive.remote())
+
+
+@pytest.mark.parametrize(
+    "ray_instance",
+    [
+        {
+            "RAY_SERVE_HTTP_PROXY_CALLBACK_IMPORT_PATH": RETURN_BAD_OBJECTS_IMPORT_PATH,
+        },
+    ],
+    indirect=True,
+)
+def test_http_proxy_return_aribitary_objects(ray_instance):
+    """Test invalid callback path in http proxy"""
+
+    actor_def = ray.serve._private.http_proxy.HTTPProxyActor
+    handle = actor_def.remote(
+        host="http_proxy",
+        port=123,
+        root_path="/",
+        controller_name="controller",
+        node_ip_address="127.0.0.1",
+    )
```

Comment:
can we test the actual user-facing behavior instead of this?

i.e., check that the proxy status goes to `UNHEALTHY` and that error message shows up in the controller log

Reply:
if checking controller log is hard/hacky that's OK

Reply:
It is doable, added a test check controller log and make sure http proxy is set to be unhealthy.

Reply:
(ideally we should have a separate test for http proxy, and make sure any exception from constructor can lead to http proxy state to be `UNHEALTHY`.)